soci/gzip_indexer.go

```diff
@@ -0,0 +1,147 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package soci
+
+// #cgo CFLAGS: -I${SRCDIR}/../c/
+// #cgo LDFLAGS: -L${SRCDIR}/../out -lindexer -lz
+// #include "indexer.h"
+// #include <stdlib.h>
+// #include <stdint.h>
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type GzipIndexer struct {
+	index *C.struct_gzip_index
+}
+
+// GenerateIndex wraps `C.generate_index` and should be used for index generation instead of relying on C code.
+// Returns the byte slice containing the `gzip_index`.
+func GenerateIndex(gzipFile string, spanSize int64) ([]byte, error) {
```

Comment:
The API is confusing to me. If I want to just a create a `GzipIndex`, I have to call this function to a byte array and then call `NewGzipIndexer` to convert it back to an index. 

Maybe have `GenerateIndex` just return a `GzipIndex` and have a `func (g *GzipIndex) Bytes() []byte`. This would be useful in cases where people want to just create the index and use it directly without converting into byte data.

Reply:
There are two use cases. Generate the index for Ztoc - you pass the file name and span size and get the index. Another use case, you have Ztoc and create Index from byte slice. And snapshotter anticipates using the latter use case within span manager. `NewGzipIndex` will be using ztoc-s IndexByteData, and `NewGzipIndexFromFile` will using tar.gz file to create `*GzipIndex`